### PR TITLE
feat(lambda): add support for auto-instrumentation with ADOT Lambda layer

### PR DESCRIPTION
**Pull request by huyphan**
AWS Distro for OpenTelemetry (ADOT) layers provide OpenTelemetry agent with an out-of-the-box configuration for AWS Lambda and AWS X-Ray. For the popular runtimes (Java, Python, Node), it provides automatic instrumentation without code change. This commit allows CDK users to easily add an ADOT Lambda layer to their Lambda function with the new prop `adotInstrumentationConfig`. For example:

```ts
const fn = new lambda.Function(this, 'MyFunction', {
  ...
  adotInstrumentation: {
    layerVersion: AdotLayerVersion.fromJavaScriptSdkLayerVersion(AdotLambdaLayerJavaScriptSdkVersion.V1_7_0),
    execWrapper: AdotLambdaExecWrapper.REGULAR_HANDLER,
  }
  ...
});
```

To support this feature, we provides the following a new classes, enum, and helper functions to users:
* `AdotLambdaExecWrapper` is the enum of all possible wrapper script that's provided by ADOT Lambda layer and must be set to the `AWS_LAMBDA_EXEC_WRAPPER` environment variable
* `AdotLambdaLayer*Version` are five classes representing five different Lambda layer types offered by ADOT. 
* `AdotLayerVersion.from*LayerVersion` are five helper functions that return the layer ARN for the Lambda function during synthesis time. 

If users want to retrieve the ARN independently from enabling ADOT in the Lambda function, they can use the following pattern:

```ts
declare const fn: lambda.Function;
const layerArn = lambda.AdotLambdaLayerJavaSdkVersion.V1_19_0.layerArn(fn.stack, fn.architecture);
```

----

### All Submissions:

* [X] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [X] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [X] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/core/lib/stack.ts [hunk 1]
```diff
@@ -960,7 +960,10 @@ export class Stack extends Construct implements ITaggable {
       throw new Error(`Context value '${cxapi.TARGET_PARTITIONS}' should be a list of strings, got: ${JSON.stringify(partitions)}`);
     }
 
-    const lookupMap = partitions ? RegionInfo.limitedRegionMap(factName, partitions) : RegionInfo.regionMap(factName);
+    const lookupMap =
+      partitions !== undefined && partitions !== 'undefined'
+        ? RegionInfo.limitedRegionMap(factName, partitions)
+        : RegionInfo.regionMap(factName);
```
Comment (huyphan) [inline packages/@aws-cdk/core/lib/stack.ts:966]: These changes belong to a commit that was submitted in a separate PR: https://github.com/aws/aws-cdk/pull/23023

I have to rebase-on-top/include this commit in this PR because without it, the added integration test will fail.

**Comment by bryan-aguilar** [comment]
I may have missed it but how does the lookup resolve the layer version. For example `arn:aws:lambda:ca-central-1:901920570463:layer:aws-otel-java-wrapper-arm64-ver-1-19-0:2`. Will the lookup always pull in the latest layer version which in this case is `:2`?

**Comment by bryan-aguilar** [comment]
Ahh, the answer lies in `fact-tables.ts`. GitHub hid that diff from me and I didn't expand it on the first pass. My question answered itself.

**Review by corymhall — CHANGES_REQUESTED** [review]
Thanks for the PR! It looks like there are a couple of things that we need to
add.

1. From the docs it looks like enabling the layer requires you to also:
	- Enable active tracing
	- Set the `AWS_LAMBDA_EXEC_WRAPPER` env variable
	- Go & .NET have some special requirements. We should handle what we can
	  (e.g. check that Go lambdas use provided.al2) and point users to the docs
	  for info on how to configure the rest

2. Version support. I know we use `LATEST` as defaults elsewhere, but we are
   trying to get away from that. What if we turned the `AdotLambdaLayerType`
   into an [enum-like class](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md#enums)
   and allowed the user to specify the version? Something like:

```ts
export class AdotJavaVersion {
  /**
   * Will always use the latest layer version. New versions could
   * introduce incompatable changes. Make sure to test new versions.
   */
  public static LATEST = new AdotJavaVersion('LATEST');

  public static 1_19_0_1 = new AdotJavaVersion('1-19-0:1');

  constructor(public readonly version: string) {}
}
export class AdotLambdaLayerType {
  public static fromJavaVersion(version: AdotJavaVersion, architecture?: Architecture): AdotLambdaLayerType {
    ...
  }
}
```

The user can still use `LATEST` if they want to, but they need to opt-in.

File: packages/@aws-cdk/region-info/build-tools/fact-tables.ts [hunk 2]
```diff
@@ -500,3 +500,218 @@ export const FIREHOSE_CIDR_BLOCKS: { [region: string]: string } = {
   'us-west-1': '13.57.135.192',
   'us-west-2': '52.89.255.224',
 };
+
+const ADOT_LAMBDA_LAYER_JAVA_SDK_ARNS: { [version: string]: { [arch: string]: { [region: string]: string } } } = {
```
Comment (rapphil) [inline packages/@aws-cdk/region-info/build-tools/fact-tables.ts:504]: Do we need to hard code the account id multiple times? Can't we generate all the permutations of region/accountid/language programmatically?

I know that you are following what was in the repo before, but I'm curious if we cannot consider this. It is very easy for a typo to happen here.

**Comment by huyphan** [comment]
>

**Comment by huyphan** [comment]
@corymhall thanks for the feedback! 

> 1. From the docs it looks like enabling the layer requires you to also:
>    
>    * Enable active tracing
>    * Set the `AWS_LAMBDA_EXEC_WRAPPER` env variable
>    * Go & .NET have some special requirements. We should handle what we can
>      (e.g. check that Go lambdas use provided.al2) and point users to the docs
>      for info on how to configure the rest

I'd describe that in a slightly different way: enabling auto-instrumentation requires you to add the layer and those extra steps. I'm saying that because one can legitimately acquire an ADOT layer just to consume the libraries that it offers and not to enable instrumentation (thus does not need to perform those extra steps). 

I can add support for those steps if we prefer coupling this way. I would need to change the API from `function.addAdotLayer()` to `function.addAutoInstrumentation()`  or `function.addAdotInstrumentation()`. 

> 2. Version support. I know we use `LATEST` as defaults elsewhere, but we are
>    trying to get away from that. What if we turned the `AdotLambdaLayerType`
>    into an [enum-like class](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md#enums)
>    and allowed the user to specify the version? Something like:
> 
> ```ts
> export class AdotJavaVersion {
>   /**
>    * Will always use the latest layer version. New versions could
>    * introduce incompatable changes. Make sure to test new versions.
>    */
>   public static LATEST = new AdotJavaVersion('LATEST');
> 
>   public static 1_19_0_1 = new AdotJavaVersion('1-19-0:1');
> 
>   constructor(public readonly version: string) {}
> }
> export class AdotLambdaLayerType {
>   public static fromJavaVersion(version: AdotJavaVersion, architecture?: Architecture): AdotLambdaLayerType {
>     ...
>   }
> }
> ```
> 
> The user can still use `LATEST` if they want to, but they need to opt-in.

I saw that we did this with CloudWatch Insights and I tried to follow the pattern. But because we have 5 different Lambda layer types and each of them has their own version line, the code quickly got messy. The fact that ADOT has inconsistent layer revisions for the same layer version makes the problem worse. For example:
1. The Java layer uses the version `1-19-0:1` for all regions/architectures but `1-19-0:2` for `arm64` in `ca-central-
2.1` (https://aws-otel.github.io/docs/getting-started/lambda/lambda-java)
2. The JS layer uses the version `1-7-0:2` for some regions  and `1-7-0:1` in some regions for `arm64` (https://aws-otel.github.io/docs/getting-started/lambda/lambda-js).

**Comment by corymhall** [comment]
> I'd describe that in a slightly different way: enabling auto-instrumentation requires you to add the layer and those extra steps. I emphasized that slight difference because one can legitimately acquire an ADOT layer just to consume the libraries that it offers and not to enable instrumentation (thus does not need to perform those extra steps). 

Ok I didn't realize that was a valid use case. Could we do something similar to
CloudWatch Insights/my example class and just have the auto-instrumentation
setup be an extra flag?

> The fact that ADOT has inconsistent layer revisions for the same layer version makes the problem worse. For example:

That is unfortunate for the user experience, but we need to make the user
specify the version. If the ADOT team has built the versioning in a way that the
user is required to know the version per region then the only thing we can do is
potentially add some validation to check that the user entered a valid version
for their region.

**Comment by huyphan** [comment]
@corymhall the latest revision should address your comments. Few notes:
* Instead of using an extra prop in `FunctionProps`, I chose to add this support via a helper function since it allows consumers to conditionally enable instrumentation in a cleaner way like this:
```
if (!isProduction) {
  fn.addAdotInstrumentation({ .. });
}
```
* For the names of the static properties representing the versions, I use the pattern `V1_19_0` instead of `1_19_0` for the same reason as why we did it with [`aws-neptune.EngineVersion`](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/aws-neptune/lib/cluster.ts#L22)

File: packages/@aws-cdk/aws-lambda/lib/adot-layers.ts [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+  /**
+   * The version of the Lambda layer
+   */
+  readonly version:
+  | AdotLambdaLayerJavaSdkVersion
+  | AdotLambdaLayerJavaAutoInstrumentationVersion
+  | AdotLambdaLayerJavaScriptSdkVersion
+  | AdotLambdaLayerPythonSdkVersion
+  | AdotLambdaLayerGenericVersion;
```
Comment (corymhall) [inline packages/@aws-cdk/aws-lambda/lib/adot-layers.ts:152]: jsii doesn't allow union types (I'm not sure why the build is passing actually).
What if we combined the type with the version?

```ts
export abstract class AdotLambdaLayerType {
  public static fromJavaSdkVersion(version: AdotLambdaLayerJavaSdkVersion, execWrapper: string): AdotLambdaLayerType {
    return new JavaSdkVersionLayer({
      version: version.version,
      type: AdotLayerType.JAVA_SDK,
    });
  }

  public abstract bind(scope: Construct, handler: IFunction, options?: LayerBindOptions): AdotLambdaLayerConfig;
}

class JavaSdkVersionLayer extends AdotLambdaLayerType {
  constructor(private readonly options: JavaSdkVersionLayerOptions) {
    super();
  }

  public bind(scope: Construct, handler: IFunction, options: LayerBindOptions = {}): AdotLambdaLayerConfig {
    // can do the individual runtime checks here, just an example
    if (handler.runtime === Runtime.JAVA_8) {
      throw new Error(...);
    }	
    return {
      version: this.options.version,
      architecture: handler.architecture,
      type: this.options.type,
    };
  }
}
```

Then in the addAdotLayer method you can just do
```ts
const config = props.type.bind(this, this);
AdotLambdaLayer.fromLayerAttributes(this, 'Layer', config);
```

The user would do something like

```ts
handler.addAdotLayer({
  type: AdotLambdaLayerType.fromJavaSdkVersion(AdotLambdaLayerJavaSdkVersion.V1_9_1),
})
```

File: packages/@aws-cdk/aws-lambda/lib/adot-layers.ts [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+export interface AdotInstrumentationProps {
+  /**
+   * The type of the ADOT Lambda layer.
+   */
+  readonly type: AdotLambdaLayerType;
+
+  /**
+   * The startup script to run, see ADOT documentation to pick the right script for your use case: https://aws-otel.github.io/docs/getting-started/lambda
+   */
+  readonly execWrapper: string;
```
Comment (corymhall) [inline packages/@aws-cdk/aws-lambda/lib/adot-layers.ts:142]: Can we make this an enum? Looking at the docs it looks like there are only a
couple of valid options.

```ts
class AdotExecWrapper {
  public static HANDLER = '/opt/otel-handler',
  public static PROXY_HANDLER = '/opt/otel-proxy-handler',
  public static STREAM_HANDLER = '/opt/otel-stream-handler',

  public static customHandler(handler: string): string {
    return handler;
  }
}
```

**Review by corymhall — DISMISSED** [review]
Looking good! Couple more comments.

File: packages/@aws-cdk/aws-lambda/lib/adot-layers.ts (hunk 3 above)
Comment (huyphan) [reply packages/@aws-cdk/aws-lambda/lib/adot-layers.ts]: JSII does allow union types, it just does not work in other languages the same way as it does in TypeScript.

Since all those `*Version` classes inherits a base class, how about we just use that base class instead? My latest revision demonstrates the idea. But if you still prefer to have a specific function for each ADOT layer type (which I honestly want to avoid), I can do that.

File: packages/@aws-cdk/aws-lambda/lib/adot-layers.ts (hunk 4 above)
Comment (huyphan) [reply packages/@aws-cdk/aws-lambda/lib/adot-layers.ts]: We can. I've updated this in the latest revision. I chose to use a real enum instead of a enum-like class to make the interface simpler. If users really want to use a custom handler, they can always override the `AWS_LAMBDA_EXEC_WRAPPER` environment variable themselves with `fn.addEnvironment();`.

This is a two-way door as we can still switch to the enum-like class option later (if necessary).

**Review by corymhall — DISMISSED** [review]

**Comment by huyphan** [comment]
@corymhall 

> Why do you want to avoid the specific functions per type?

Too much boilerplate code and extra interfaces are the main reason. For each lambda layer type there will be one new `fromX` function. Since those functions receive more than one attributes, each requires its own props/attrs interface. 

I agree that customer experience will be better with that. Let me give that a try anyway.

**Comment by huyphan** [comment]
> With the currently implementation we need to input the same type of information multiple places,
and make sure that the user has input the correct combination.

Note that the implementation you suggested does not stop users from inputting the wrong combination of type and version. For example, with these declarations:

```typescript
class AdotLambdaLayerJavaSdkVersion {
  public static readonly V1_19_0 = new AdotLambdaLayerJavaSdkVersion('1.19.0');
  protected constructor(public readonly version: string) {}
}

class AdotLambdaLayerPythonSdkVersion {
  public static readonly V1_13_0 = new AdotLambdaLayerPythonSdkVersion('1.13.0');
  protected constructor(public readonly version: string) {}
}

function createJavaLayer(version: AdotLambdaLayerJavaSdkVersion) {
  console.log("Run just fine");
}
```

You would think that this function invocation won't be allowed, but it's actually accepted thanks to TypeScript being duck-typing:

```typescript
createJavaLayer(AdotLambdaLayerPythonSdkVersion.V1_13_0);
```

**Comment by huyphan** [comment]
The latest revision is ready for review now. It should be very close to what you suggested in term of customer experience. Also, I use the "branding" trick to work around the issue of structural typing.

**Comment by huyphan** [comment]
The module `lambda-insights.ts` gave me some good hints so I decided to scratch the current design and re-structured the `adot-layers.ts` module a bit. I think I'm happy with the latest code. The usage, as of now, is:

```
fn.addAdotInstrumentation({
  layerVersion: AdotLambdaLayerJavaSdkVersion.V1_19_0,
  execWrapper: AdotLambdaExecWrapper.REGULAR_HANDLER,
});
``` 

I'm going to call it a day now. Let me know if you have any question there.

File: packages/@aws-cdk/aws-lambda/lib/adot-layers.ts [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+
+/**
+ * The collection of versions of the ADOT Lambda Layer for JavaScript SDK
+ */
+export class AdotLambdaLayerJavaScriptSdkVersion extends AdotLambdaLayerVersion {
+  /**
+   * Will always use the latest layer version. New versions could
+   * introduce incompatible changes. Make sure to test new versions.
+   */
+  public static readonly LATEST = new AdotLambdaLayerJavaScriptSdkVersion('1.7.0');
```
Comment (bryan-aguilar) [inline packages/@aws-cdk/aws-lambda/lib/adot-layers.ts:221]: Is there a pattern here that requires the use of a `LATEST` tag? The lambda layers don't really currently push any reference to `latest` and this would be the first introduction of the term in relation to the ADOT lambda layers. I worry that it could be confusing to end-users.

Comment (huyphan) [reply packages/@aws-cdk/aws-lambda/lib/adot-layers.ts:221]: If I update the wording as below, would that make more sense?

```
Use the latest layer version available within this CDK version ...
```

The `LATEST` version would be helpful for those who want to model a ADOT layer update via CDK library update.

Comment (huyphan) [reply packages/@aws-cdk/aws-lambda/lib/adot-layers.ts:221]: I went ahead and updated the comment. Please see the latest revision.

**Review by corymhall — DISMISSED** [review]

File: packages/@aws-cdk/aws-lambda/README.md [hunk 6]
```diff
@@ -764,6 +764,40 @@ const fn = new lambda.Function(this, 'MyFunction', {
...
+import {
+  AdotLambdaExecWrapper,
+  AdotLambdaLayerJavaScriptSdkVersion,
+} from 'aws-cdk-lib/aws-lambda';
+
+const fn = new lambda.Function(this, 'MyFunction', {
+  runtime: lambda.Runtime.NODEJS_18_X,
+  handler: 'index.handler',
+  code: lambda.Code.fromInline('exports.handler = function(event, ctx, cb) { return cb(null, "hi"); }'),
+  adotInstrumentationConfig: {
```
[the remark this answers is not in the document]
Comment (huyphan) [reply packages/@aws-cdk/aws-lambda/README.md:783]: We should avoid confusing users about two slightly different concepts here: *ADOT instrumentation* and *ADOT Lambda layer*. Enabling ADOT instrumentation means two independent activities:
1. Adding an ADOT Lambda layer to the Lambda function.
2. Configuring the Lambda function to use an exec wrapper that's compatible with ADOT.

So setting a value for the exec wrapper means configuring the instrumentation, not configuring the layer.  I'm afraid that the proposed usage suggests the opposite. 

What do you think about this?

```ts
const fn = new lambda.Function(this, 'MyFunction', {
  ...
  adotInstrumentationConfig: AdotInstrumentationConfig.forJavaScriptSdk(
    AdotLambdaLayerJavaScriptSdkVersion.V1_7_0,
    AdotLambdaExecWrapper.REGULAR_HANDLER, // optional, defaulted to REGULAR_HANDLER
  ),
  ...
});
```
One thing I don't like about this approach though, is that we will have to introduce 5 more helper functions under `AdotLambdaConfig`. But I'm happy to go with it if it's right for users.

[the remark this answers is not in the document]
Comment (corymhall) [reply packages/@aws-cdk/aws-lambda/README.md:783]: Ok sure lets go with `adotInstrumentation` then (drop the `Config`).

[the remark this answers is not in the document]
Comment (huyphan) [reply packages/@aws-cdk/aws-lambda/README.md]: Sure! I've added a new commit to address this.

**Review by corymhall — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-lambda/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (huyphan) [reply packages/@aws-cdk/aws-lambda/README.md]: I'm a bit skeptical with that experience. When I tried that new change in the code, and write this line as a user:

```ts
layerVersion: lambda.AdotLayerVersion.fromJavaSdkLayerVersion(lambda.AdotLambdaLayerJavaSdkVersion.LATEST),

```

... two things stood out for me:

1. When I typed up to `layerVersion: lambda.`, the auto-complete feature suggested me a bunch of options. The SDK-specific options are showed up at the top and `AdotLayerVersion`, which is the only one I can use, is near to the bottom. Most of the times, users know what SDK they need, so they would tend to choose those SDK-specific option first and get stuck. 
2. I have to type the same Layer type (i.e., `JavaSdk` ) twice on that line. 

---

Given those concerns, if you think they are minor enough, I'm happy to update the code to align with that experience.

[the remark this answers is not in the document]
Comment (corymhall) [reply packages/@aws-cdk/aws-lambda/README.md]: But my editor will show me that `layerVersion` takes an `AdotLayerVersion` type so I can just type that in and when I get to the `.` it will show me all the options. The alternative that you propose is that it would show me that it takes a type of `AdotLambdaLayerVersion` and I have to just know which other classes implement that. Nothing  in the editor will help me out.


> I have to type the same Layer type (i.e., JavaSdk ) twice on that line.

Yeah I get that, but that's just the way the versions work right? If every language had the same versions then this would be easier. We use this model elsewhere, for example the [rds.DatabaseCluster.engine](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_rds.DatabaseCluster.html#engine)

[the remark this answers is not in the document]
Comment (huyphan) [reply packages/@aws-cdk/aws-lambda/README.md]: I've updated the code ([the diff on top of old commits](https://github.com/aws/aws-cdk/commit/d20be2ab05d74db6ba1c7e8cc006fd696eda1573#diff-23d1a5e447cf921c3bb1b55507c661f44dd53e5317c0a4d35ea1ea6ed5c9cc6d)) to make the UX aligned with what you proposed. The code is a bit different from your gist since I want to avoid exposing irrelevant props to users. Let me know if you need me to change anything else.

**Comment by huyphan** [comment]
There was a small issue with the README change in my last commit. But it's all sorted now.

**Review by corymhall — APPROVED** [review]
@huyphan looks great! thanks for being patient and sticking with me on this one! 🚀
